src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,355 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Threading.Channels;
+using Aspire.Dashboard.Utils;
+using Aspire.V1;
+using Grpc.Core;
+using Grpc.Net.Client;
+using Grpc.Net.Client.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Dashboard.Model;
+
+/// <summary>
+/// Implements gRPC client that communicates with a resource server, populating data for the dashboard.
+/// </summary>
+/// <remarks>
+/// An instance of this type is created per service call, so this class should not hold onto any state
+/// expected to live longer than a single RPC request. In the case of streaming requests, the instance
+/// lives until the stream is closed.
+/// </remarks>
+internal sealed class DashboardClient : IDashboardClient
+{
+    private const string DashboardServiceUrlVariableName = "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL";
+    private const string DashboardServiceUrlDefaultValue = "http://localhost:18999";
+
+    private readonly Dictionary<string, ResourceViewModel> _resourceByName = new(StringComparers.ResourceName);
+    private readonly CancellationTokenSource _cts = new();
+    private readonly TaskCompletionSource _whenConnected = new();
+    private readonly object _lock = new();
+    private readonly ILogger<DashboardClient> _logger;
+
+    private ImmutableHashSet<Channel<ResourceViewModelChange>> _outgoingChannels = [];
+    private string? _applicationName;
+
+    private const int StateNone = 0;
+    private const int StateInitialized = 1;
+    private const int StateDisposed = 2;
+    private int _state;
+    private readonly GrpcChannel _channel;
+    private readonly DashboardService.DashboardServiceClient _client;
+    private Task? _connection;
+
+    public DashboardClient(ILogger<DashboardClient> logger)
+    {
+        _logger = logger;
```

Comment:
Use logger factory in ctor so it can be passed to GrpcChannel later.
```suggestion
    public DashboardClient(ILoggerFactory loggerFactory)
    {
        _logger = loggerFactory.CreateLogger<DashboardClient>();
```
Then set the logger factory when the `GrpcChannel` is created so its activity is logged by the app: https://learn.microsoft.com/en-us/aspnet/core/grpc/diagnostics?view=aspnetcore-8.0#grpc-client-logging